pyblazing/pyblazing/apiv2/context.py

```diff
@@ -681,36 +681,57 @@ def adjust_due_missing_rowgroups(metadata, files):
 def distributed_initialize_server_directory(client, dir_path):
 
     # lets make host_list which is a list of all the unique hosts.
-    # This way we do the logging folder creation only once per host (server)
+    # this way we do the logging folder creation only once per host (server)
+    # or for each worker whenever run from different directories
     all_items = client.scheduler_info()["workers"].items()
-    host_list = list(set([value["host"] for key, value in all_items]))
-    initialized = {}
-    for host in host_list:
-        initialized[host] = False
 
     dask_futures = []
-    for worker, worker_info in client.scheduler_info()["workers"].items():
-        if not initialized[worker_info["host"]]:
-            dask_futures.append(
-                client.submit(initialize_server_directory, dir_path, workers=[worker])
-            )
-            initialized[worker_info["host"]] = True
+    i = 0  # this extra index is to avoid scheduler optimization
+    for worker, worker_info in all_items:
+        dask_futures.append(
+            client.submit(get_current_directory_path, i, workers=[worker])
+        )
+        i = i + 1
+
+    current_dir_hosts = client.gather(dask_futures)
+
+    # lets map the working directories shared by several workers,
+    # in such a way that only one worker will create the logging folder
+    map_host_path = {}
+    for path, worker in zip(
+        current_dir_hosts, [worker for worker, worker_info in all_items]
+    ):
+        map_host_path[path] = worker
```

Comment:
I dont think this will work. If all workers have the same current working directory, then your `map_host_path` will only have one entry, which means only one of the workers will get initialized

Reply:
Additionally, this is only necessary if clients have different CWD AND if the `dir_path` is a relative path. If its an absolute path, it does not matter what the CWD is. Which means what you need to get is a set of  unique worker IP and actual resolved `dir_path`. By resolved, i mean if its a relative path, then "resolve" it against the CWD, if its absolute, then its already resolved.

Reply:
You're right @williamBlazing, I didn't consider these scenarios. Thanks, I just improved the solution.